### PR TITLE
sql: infrastructure and fixes for preserving statement for retries

### LINKED ISSUE #4472: sql: server side transaction retries are broken for groups, subqueries
The various visitors that mutate expressions leak these modifications to the next retry of the transaction. We need a more robust scheme of mutating expressions.

I verified that we have a problem with subqueries - when retrying `SELECT k FROM T where k NOT IN (SELECT k from T)` we are actually running `SELECT k FROM T where k NOT IN (1, 2, 3..)`. The values we see in the table could have changed. (It's hard to actually make this return an incorrect result in on a single node though).

### PR DESCRIPTION
**Pull request by RaduBerinde**
These changes implement the Walk solution we discussed over email and fix #4472.

I split the changes into a few commits, it should be easier to review separately.

@mjibson the third commit touches on some of the returning stuff, it would be great if you could take a look.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4609)

### CODE REVIEW

**Comment by tamird** [comment]
LGTM. My review was pretty quick though, I didn't read every line, so perhaps wait for @mjibson and @andreimatei's reviews

---

Reviewed 8 of 8 files at r1, 8 of 8 files at r2, 6 of 6 files at r3, 10 of 10 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/executor.go, line 42 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFPiMFOqhf19sKf5rr:-KBFPiMFOqhf19sKf5rs:-1252689397) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fc875915d54fbf32a0a8b77073e1d10505ca8e/sql/executor.go#L42)):*
should this be on just for tests?

---

*[sql/parser/expr.go, line 288 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFPlyS5IbatUv-VEdM:-KBFPlyS5IbatUv-VEdN:-361868826) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fc875915d54fbf32a0a8b77073e1d10505ca8e/sql/parser/expr.go#L288)):*
should this be opted out of outside of tests?

---

*[sql/parser/returning.go, line 28 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFPNdxfP8XS2l1WiJK:-KBFPNdxfP8XS2l1WiJL:1840735822) ([raw file](https://github.com/cockroachdb/cockroach/blob/490072da268755d89b71470a7a844fa4075179b5/sql/parser/returning.go#L28)):*
was this intended?

---

*[sql/parser/walk.go, line 80 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFOntw4T7Kb7VxxZ8w:-KBFOntw4T7Kb7VxxZ8x:236679929) ([raw file](https://github.com/cockroachdb/cockroach/blob/e34859c9ff689247e28f191b719c52502dcac2aa/sql/parser/walk.go#L80)):*
does `CopyNode` need to be part of the Expr interface?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-:-KBFPtwI2uRCWG6mPrCm:1277105826)*

**Comment by RaduBerinde** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/executor.go, line 42 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFPiMFOqhf19sKf5rr:-KBFQgNvRdzruxawnokF:-1395325497) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fc875915d54fbf32a0a8b77073e1d10505ca8e/sql/executor.go#L42)):*
TFTR! I am not sure, I wanted to see what people think. Given how many separate instances of in-place modification there were, I think I would like to keep it on at least for a while.

---

*[sql/parser/expr.go, line 288 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFPlyS5IbatUv-VEdM:-KBFQTTOKcRu9ykKwiHR:-978738022) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fc875915d54fbf32a0a8b77073e1d10505ca8e/sql/parser/expr.go#L288)):*
I think it's useful enough to keep for the improvement in error messages.

---

*[sql/parser/returning.go, line 28 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFPNdxfP8XS2l1WiJK:-KBFRCHLZF2UQWTISnwB:307895178) ([raw file](https://github.com/cockroachdb/cockroach/blob/490072da268755d89b71470a7a844fa4075179b5/sql/parser/returning.go#L28)):*
yes, `SelectExprs` starts with a space.

---

*[sql/parser/walk.go, line 80 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFOntw4T7Kb7VxxZ8w:-KBFRFeM6M8lYsp7jTEq:-581392295) ([raw file](https://github.com/cockroachdb/cockroach/blob/e34859c9ff689247e28f191b719c52502dcac2aa/sql/parser/walk.go#L80)):*
No, I only defined these as a helper function for a few non-trivial types (and it returns the particular type not a generic `Expr`, which would not be possible if it was in the interface). I exported them because they can be useful, e.g. when writing visitors.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/executor.go, line 42 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFPiMFOqhf19sKf5rr:-KBFSYvuqSqF7H53Mc3k:567654103) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fc875915d54fbf32a0a8b77073e1d10505ca8e/sql/executor.go#L42)):*
OK. Can you add a TODO?

---

*[sql/parser/expr.go, line 288 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFPlyS5IbatUv-VEdM:-KBFSk-d7pShEyEVW2ai:211439921) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fc875915d54fbf32a0a8b77073e1d10505ca8e/sql/parser/expr.go#L288)):*
Oh, yes, ignore me.

---

*[sql/parser/returning.go, line 28 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFPNdxfP8XS2l1WiJK:-KBFT-Jb59Jq4p_zNwPW:-798713675) ([raw file](https://github.com/cockroachdb/cockroach/blob/490072da268755d89b71470a7a844fa4075179b5/sql/parser/returning.go#L28)):*
Heh, I guess we have no test for this in expr_test.go. Mind adding one?

---

*[sql/parser/walk.go, line 80 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFOntw4T7Kb7VxxZ8w:-KBFT6HlEyQJ1iMKpdr1:676269983) ([raw file](https://github.com/cockroachdb/cockroach/blob/e34859c9ff689247e28f191b719c52502dcac2aa/sql/parser/walk.go#L80)):*
:+1:

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609)*

**Comment by RaduBerinde** [comment]
*[sql/parser/returning.go, line 28 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFPNdxfP8XS2l1WiJK:-KBFTk3nHB99UWTbX-Bb:1341788455) ([raw file](https://github.com/cockroachdb/cockroach/blob/490072da268755d89b71470a7a844fa4075179b5/sql/parser/returning.go#L28)):*
Sure. I will add in parse_test.go

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609)*

**Comment by mjibson** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/returning.go, line 29 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFAH1IzVlETL586crn:-KBFAH1IzVlETL586cro:291402798) ([raw file](https://github.com/cockroachdb/cockroach/blob/637ba14dbcde5467e4775fedb4fdf562b3983a37/sql/returning.go#L29)):*
This is not idiomatic Go. How about something like `rh, err := newReturning(...)` is more common.

---

*[sql/returning.go, line 85 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFAkbn4cXTmGQHKeo2:-KBFAkbn4cXTmGQHKeo3:1333554230) ([raw file](https://github.com/cockroachdb/cockroach/blob/637ba14dbcde5467e4775fedb4fdf562b3983a37/sql/returning.go#L85)):*
Since this method doesn't have a side-effect, the name `finalize` is misleading. Is there any problem just using `rh.values` directly in the caller instead of a getter method?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609)*

**Comment by mjibson** [comment]
Returning stuff LGTM

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-:-KBJTK7S6Rs-f_0HxYEG:521631923)*

File: sql/parser/walk.go [hunk 1]
```diff
@@ -16,206 +16,701 @@
...
 // WalkExpr traverses the nodes in an expression.
-func WalkExpr(v Visitor, expr Expr) Expr {
-	v, expr = v.Visit(expr, true)
-	if v == nil {
-		return expr
+func WalkExpr(v Visitor, expr Expr) (new Expr, changed bool) {
+	recurse, new := v.VisitPre(expr)
+
+	if recurse {
+		new = new.Walk(v)
```
Comment (mjibson) [inline sql/parser/walk.go]: `new` is a builtin. Could we use another name to not shadow it?

File: sql/parser/deep_copy_test.go [hunk 2]
```diff
@@ -1,371 +0,0 @@
-// Copyright 2016 The Cockroach Authors.
```
Comment (mjibson) [inline sql/parser/deep_copy_test.go]: Do we not need to migrate this testing to the new WalkStmt stuff?

**Comment by RaduBerinde** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/parser/walk.go, line 387 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBF4EjbOObbcaaqZrWx-r6-387:-KBJgu-oA33jvbTbtH1D:-992469219) ([raw file](https://github.com/cockroachdb/cockroach/blob/053f2ce5780af9705d5b371de3641d59e357576b/sql/parser/walk.go#L387)):*
Ha, it's even syntax-highlighted here. Will fix.

---

*[sql/returning.go, line 29 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFAH1IzVlETL586crn:-KBJhM-oNWMvmLn5veqY:1855604471) ([raw file](https://github.com/cockroachdb/cockroach/blob/637ba14dbcde5467e4775fedb4fdf562b3983a37/sql/returning.go#L29)):*
I will do that. I think I did it this way so we wouldn't need to allocate it but it's a pretty dumb thing to optimize here.

---

*[sql/returning.go, line 85 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFAkbn4cXTmGQHKeo2:-KBJhSaWp3vn0UKNOl85:-291629499) ([raw file](https://github.com/cockroachdb/cockroach/blob/637ba14dbcde5467e4775fedb4fdf562b3983a37/sql/returning.go#L85)):*
I figured it would be nice to give the class freedom to do some final processing (and it's clearer how the results are used by just reading that file). I will keep the function but change it to `getResults`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609)*

File: sql/parser/deep_copy_test.go (hunk 2 above)
Comment (RaduBerinde) [inline sql/parser/deep_copy_test.go]: We don't need the deep copy stuff anymore. It had kind of a faulty semantic ("deep" meant up to Expr leaves but did not deep copy things like `DTuple` or `*DDecimal`). There are existing tests for Walk and Normalize, and I added a check that the original expression is not modified to those.

**Comment by andreimatei** [comment]
LGTM

---

Reviewed 5 of 8 files at r1, 3 of 8 files at r2, 2 of 6 files at r3.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 42 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJlT1aQ0PIKj3UBYrL:-KBJlT1aQ0PIKj3UBYrM:394763180) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fc875915d54fbf32a0a8b77073e1d10505ca8e/sql/executor.go#L42)):*
well what's the point of this? I think we should either get rid of it (always do the check), or move it to a pair of special go files that init constants, guarded by debug/non-debug builds?

---

*[sql/parser/normalize_test.go, line 30 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJSP3qEYEb2HKW_MtX:-KBJSP3qEYEb2HKW_MtY:712607617) ([raw file](https://github.com/cockroachdb/cockroach/blob/e34859c9ff689247e28f191b719c52502dcac2aa/sql/parser/normalize_test.go#L30)):*
what's up with this?

---

*[sql/parser/walk_test.go, line 145 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJXZdckBjMHycL0TdT:-KBJXZddP5U22cyTnFnm:-703063157) ([raw file](https://github.com/cockroachdb/cockroach/blob/65d5c4845cf653bb8e2f1c1b0e5bd872b617e6b9/sql/parser/walk_test.go#L145)):*
Please add a comment here that this is where the visiting happens. Took me a bit to figure out if this test does anything.

---

*[sql/parser/walk_test.go, line 159 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJWnnqfJ1Ivwmex0V6:-KBJWnnqfJ1Ivwmex0V7:1459464489) ([raw file](https://github.com/cockroachdb/cockroach/blob/65d5c4845cf653bb8e2f1c1b0e5bd872b617e6b9/sql/parser/walk_test.go#L159)):*
This test prevents regressions of #4472 ? Maybe add a comment saying why it's important for expressions to not be modified.

Do you think this test is sufficient to prevent a regression? There's nothing about subqueryVisitor here. I guess there's also you parallel insert test.
I think it'd be good if we tested the subquery is not replaced, but I'm not sure how to do that.

---

*[sql/returning.go, line 22 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJkMFTTh__crgsV59n:-KBJkMFUyq9Qqx69wI_B:973769818) ([raw file](https://github.com/cockroachdb/cockroach/blob/637ba14dbcde5467e4775fedb4fdf562b3983a37/sql/returning.go#L22)):*
say something about how it's supposed to accumulate rows through `append`

---

*[sql/returning.go, line 24 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJjVVO62r3_niipGaV:-KBJjVVO62r3_niipGaW:-1902414518) ([raw file](https://github.com/cockroachdb/cockroach/blob/637ba14dbcde5467e4775fedb4fdf562b3983a37/sql/returning.go#L24)):*
please add a comment about this `values`being the result

---

*[sql/returning.go, line 84 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJjssmx1iKyFOpmKqw:-KBJjssnQSHbzZ3CgZj1:-1596640182) ([raw file](https://github.com/cockroachdb/cockroach/blob/637ba14dbcde5467e4775fedb4fdf562b3983a37/sql/returning.go#L84)):*
s/valuse/values

---

*[sql/select_qvalue.go, line 68 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJYjz3qzzU1_oX3ncN:-KBJYjz3qzzU1_oX3ncO:76409878) ([raw file](https://github.com/cockroachdb/cockroach/blob/490072da268755d89b71470a7a844fa4075179b5/sql/select_qvalue.go#L68)):*
how come this line went away?

---

*[sql/select_qvalue.go, line 102 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJZ5Z1esfGrdKdD_EB:-KBJZ5Z1esfGrdKdD_EC:1249221898) ([raw file](https://github.com/cockroachdb/cockroach/blob/65d5c4845cf653bb8e2f1c1b0e5bd872b617e6b9/sql/select_qvalue.go#L102)):*
so here you're changing `q`, which is against `Walk()`'s contract? But I guess this change is a special case? Is the `argsVisitor` the only one that's going to return a different expression on the line above? Maybe that visitor deserves a special comment about how it changes things in place? 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-:-KBJmBeH3DOuobYKwKia:676834233)*

**Comment by tamird** [comment]
Reviewed 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[sql/returning.go, line 29 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFAH1IzVlETL586crn:-KBJv8hVAk9D4N9gThz8:-2125235876) ([raw file](https://github.com/cockroachdb/cockroach/blob/637ba14dbcde5467e4775fedb4fdf562b3983a37/sql/returning.go#L29)):*
why don't you return a value instead of a pointer?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609)*

**Comment by RaduBerinde** [comment]
TFTR!

---

Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 42 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJlT1aQ0PIKj3UBYrL:-KBJptvTHAU_AlXMikQD:-802710878) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fc875915d54fbf32a0a8b77073e1d10505ca8e/sql/executor.go#L42)):*
I don't think we currently have debug/non-debug builds. Eventually it would be great if it was off by default but on during the tests. We should add some infrastructure for the tests to control some of these "stress" options.

---

*[sql/parser/normalize_test.go, line 30 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJSP3qEYEb2HKW_MtX:-KBJqRCHUHCL9pp9ewkD:1280081224) ([raw file](https://github.com/cockroachdb/cockroach/blob/e34859c9ff689247e28f191b719c52502dcac2aa/sql/parser/normalize_test.go#L30)):*
Good catch, I was debugging the following case and forgot to remove the comment.

---

*[sql/parser/walk_test.go, line 159 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJWnnqfJ1Ivwmex0V6:-KBJqeEQPHBtTD42LXIa:-1994975905) ([raw file](https://github.com/cockroachdb/cockroach/blob/65d5c4845cf653bb8e2f1c1b0e5bd872b617e6b9/sql/parser/walk_test.go#L159)):*
The test around `Exec` (controlled by `checkStmtStringChange`) is really the one preventing a regression.

---

*[sql/select_qvalue.go, line 68 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJYjz3qzzU1_oX3ncN:-KBJqxnQqlvcRILt2DSp:891558955) ([raw file](https://github.com/cockroachdb/cockroach/blob/490072da268755d89b71470a7a844fa4075179b5/sql/select_qvalue.go#L68)):*
It's not necessary to put the table name in the expression (which would change it and trip the check around `Exec`).

---

*[sql/select_qvalue.go, line 102 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJZ5Z1esfGrdKdD_EB:-KBJzvPePskY7exGWJ7b:-102458748) ([raw file](https://github.com/cockroachdb/cockroach/blob/65d5c4845cf653bb8e2f1c1b0e5bd872b617e6b9/sql/select_qvalue.go#L102)):*
Yeah, I will document it.. It's ok to change it in place because `qvalue` can't be part of an original expression (it's our own type).

Unfortunately at this point most visitors will have `WalkExpr` return a different expression, because of the non-pointer datums - those will look "different" as an Expr every time because a "copy" is returned. I mean that even func (b DBool) Walk() Expr { return b }`  returns an `Expr` that looks different than the `Expr` we called it with (it's a different copy which is promoted to a different object).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-:-KBKBF65MnrJmokHz_HO:-485616692)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 42 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBJlT1aQ0PIKj3UBYrL:-KBKBkI91IXCi6qAQcvO:1599452652) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6fc875915d54fbf32a0a8b77073e1d10505ca8e/sql/executor.go#L42)):*
we do have the "release" build tag. i don't think it's used anywhere.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609)*

**Comment by RaduBerinde** [comment]
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[sql/returning.go, line 29 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609#-KBFAH1IzVlETL586crn:-KBKBlDPc-XxaFwT1Npi:1085483994) ([raw file](https://github.com/cockroachdb/cockroach/blob/637ba14dbcde5467e4775fedb4fdf562b3983a37/sql/returning.go#L29)):*
Will do.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4609)*

**Pull request merged**